crates/autopilot/src/on_settlement_event.rs

```diff
@@ -306,3 +216,224 @@ impl Inner {
         }
     }
 }
+
+#[async_trait]
+impl OnSettlementEventAction for Updater {
+    async fn apply(&self) -> Result<bool> {
+        let mut ex = self
+            .db
+            .pool
+            .begin()
+            .await
+            .context("acquire DB connection")?;
+        let event = match database::settlements::get_settlement_without_auction(&mut ex)
+            .await
+            .context("get_settlement_without_auction")?
+        {
+            Some(event) => event,
+            None => return Ok(false),
+        };
+
+        let hash = H256(event.tx_hash.0);
+        tracing::debug!("updating settlement details for tx {hash:?}");
+
+        let Ok(transaction) = self.eth.transaction(hash.into()).await else {
+            tracing::warn!(?hash, "no tx found");
+            return Ok(false);
+        };
+
+        let (auction_id, auction_data) =
+            match Self::recover_auction_id_from_calldata(&mut ex, &transaction).await? {
+                AuctionIdRecoveryStatus::InvalidCalldata => {
+                    // To not get stuck on indexing the same transaction over and over again, we
+                    // insert the default auction ID (0)
+                    (Default::default(), None)
+                }
+                AuctionIdRecoveryStatus::DoNotAddAuctionData(auction_id) => (auction_id, None),
+                AuctionIdRecoveryStatus::AddAuctionData(auction_id, settlement) => (
+                    auction_id,
+                    Some(
+                        self.fetch_auction_data(hash, settlement, auction_id)
+                            .await?,
+                    ),
+                ),
+            };
+
+        let update = SettlementUpdate {
+            block_number: event.block_number,
+            log_index: event.log_index,
+            auction_id,
+            auction_data,
+        };
+
+        tracing::debug!(?hash, ?update, "updating settlement details for tx");
+
+        Postgres::update_settlement_details(&mut ex, update.clone())
+            .await
+            .with_context(|| format!("insert_settlement_details: {update:?}"))?;
+        ex.commit().await?;
+        Ok(true)
+    }
+}
+
+#[allow(dead_code)]
+pub struct CircuitBreaker {
+    eth: infra::blockchain::circuit_breaker::CircuitBreaker,
+    persistence: infra::Persistence,
+    db: Postgres,
+    solvers: Vec<eth::Address>,
+    // Registry to keep track of the already processed settlements, format (block, hash)
+    registry: Arc<Mutex<BTreeMap<i64, H256>>>,
```

Comment:
One reason I did not propose to have a cache dependent on blocks are reorganizations, that could mess it up badly. Why is having a `HashSet<hash>` dropped as an idea?

Reply:
@sunce86 By doing what I did I wanted to maximize the code performance, but if you tell block reorganizations could be a problem for it, then I can do what you originally proposed of a `dequeue` + `hashset`.